Use ZCash compressed point format (#31)
tplooker/draft-looker-bls-jose-cose: draft-ietf-cose-bls-key-representations.md

When expressing a cryptographic key for these curves in JSON Web Key (JWK) form, the following rules apply:

- The parameter "kty" MUST be present and set to "OKP".
- The parameter "crv" MUST be present and value MUST be one defined in (#curve-parameter-registration).

> - The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.

Reviewer: ```suggestion
- The parameter "x" MUST be present with its value being the base64url encoding of the compressed serialized public key curve point, as defined normatively in Appendix B of <xref target="BBS"/>.
```

Just a small nit to make clear that the encoded point is the public key

> - The parameter "d" MUST be present for private key representations with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.

Reviewer 2: Neither BBS or BLS currently define encoding for the secret key. BBS also exclusively uses big-endian representations otherwise, while today `d` is little-endian (to match the Ed25519/Ed448 native encoding). BBS needs to represent points in its protocol, but not secrets.

The updated text also misrepresents the value of `d` to be a compressed point rather than a scalar.

I would suggest preserving the existing `d` parameter text here and in COSE

Author: OK - I'll apply your suggestion later this morning

> - The parameter "d" MUST be present for private key representations with its value being the base64url encoding of the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.

Reviewer 2: ```suggestion
- The parameter "d" MUST be present for private key representations whose value MUST contain the little-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.
```

> - The parameter "d" MUST be present for private key representations whose value MUST contain the big-endian representation of the private key base64url encoded without padding as defined in [@!RFC7515] Appendix C. This parameter MUST NOT be present for public keys.

Reviewer 3: Should we also define `"d"` to be of fixed length? Similar to [RFC 7518](https://www.rfc-editor.org/rfc/rfc7518.html#section-6.2.2.1):

>The length of this octet string MUST be ceiling(log-base-2(n)/8) octets (where n is the order of the curve).

I've seen enough interop issues with presence vs. absence of leading zeroes etc (mostly in ASN.1, though) even in my short career. :slightly_smiling_face:

Author: I agree that these should be fixed length, including leading zeros. What sizes are they for each curve?

Reviewer 3: 32 bytes for [BLS12-381](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-11.html#name-bls-curves-for-the-128-bit-):

```
r: 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001
```

and 65 bytes for [BLS48-581](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-11.html#name-for-256-bit-security):

```
r: 0x2386f8a925e2885e233a9ccc1615c0d6c635387a3f0b3cbe003fad6bc972c2e6e741969d34c4c92016a85c7cd0562303c4ccbe599467c24da118a5fe6fcd671c01
```

Reviewer 2: I can generate new (matching) BLS12381G2 examples if we want to change `d` to be big-endian and fixed length.

Reviewer 3: The example `d`s are already big-endian as far as I could tell. :slightly_smiling_face:

### COSE_Key Representation

When expressing a cryptographic key for these curves in COSE_Key form, the following rules apply:

- The parameter "kty" (1) MUST be present and set to "OKP" (1).
- The parameter "crv" (-1) MUST be present and value MUST be one defined in (#curve-parameter-registration).
- The parameter "x" (-2) MUST be present with its value being the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>.

> - The parameter "d" (-4) MUST be present for private key representations with its value being the compressed serialized point format defined normatively in Appendix B of <xref target="BBS"/>. This parameter MUST NOT be present for public keys.

Reviewer 2: ```suggestion
- The parameter "d" (-4) MUST be present for private key representations whose value MUST contain the little-endian representation of the private key. This parameter MUST NOT be present for public keys.
```

## BLS12381 Key Pairs

The following examples showcase JWKs for both the G1 and G2 subgroups of the BLS12381 curve. Note, the examples also include the corresponding private key, expressed through the inclusion of the "d" parameter.

An example JWK for the BLS12381 curve where the public key is in the G1 subgroup.

```
{
  "kty": "OKP",
  "crv": "BLS12381G1",
  "x": "TBD THIS IS NOW WRONG. REPLACE IT. IJluBL6p_SQmRrZvJNa3YaJ-Wx8Im",
  "d": "TBD THIS IS NOW WRONG. REPLACE IT. I7XnEnDc"
}
```

Another example of a different JWK for the BLS12381 curve where the public key is in the G1 subgroup.

```
{
  "kty": "OKP",
  "crv": "BLS12381G1",
  "x": "TBD THIS IS NOW WRONG. REPLACE IT. _Lb-4dLu0xCaSrd0rDBMTFthd5r-2",
  "d": "TBD THIS IS NOW WRONG. REPLACE IT. uwLZ8gV4"
}
```

An example JWK for the BLS12381 curve where the public key is in the G2 subgroup.

```
{
  "kty": "OKP",
  "crv": "BLS12381G2",
  "x": "TBD THIS IS NOW WRONG. REPLACE IT. aYylVK_x0F8HhE8zk0YWiGV3CHwpQ
Ea2sH4PBZLaYCn8se-1clmCORDsKxbbw3Js_Alu4OmkV9gmbJsy1YF2rt7Vxzs6S",

>   "d": "TBD THIS IS NOW WRONG. REPLACE IT. svwjJhUM"

Reviewer 2: ```suggestion
  "x": "t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJ
DnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2",
  "d": "IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"
```

Reviewer 3: `x` and `d` are consistent in my testing.

Author: I agree with @emlun's suggestion to use the same values for the COSE keys as were used for the corresponding JOSE keys.

Reviewer 2: I would suggest doing this for the next version, and combining the two appendices so that the JWK and CWK are next to one another.

>   "d": "IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"

Reviewer 3: This `d` value is big-endian, but the above definition of `"d"` states that its "value MUST contain the little-endian representation of the private key". I think the definition should be big-endian instead (which would then agree with this example).

We can check the endianness by checking for consistency with the public key `x`. First using [@noble/curves](https://github.com/paulmillr/noble-curves) to decode the public key and confirm consistency:

```
$ node
> const { bls12_381} = require("@noble/curves/bls12-381")
> const b64u = require("base64url")
> d = bls12_381.fields.Fr.fromBytes(b64u.toBuffer("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4"))
15386317252953966243068698475713215110548824422523943950150305918933148006670n
> x = bls12_381.G2.CURVE.fromBytes(b64u.toBuffer("t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJDnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2"))
{
  x: {
    c0: 2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198n,
    c1: 3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385n
  },
  y: {
    c0: 1091225907871541971099156319884496768994573942435479350626538810779395867698340173817993335860569537230377198516854n,
    c1: 2595333569697813473526172873449378383222805423158275723970303644390236295775373772592557270513291928304374795997738n
  }
}
> bls12_381.G2.ProjectivePoint.fromPrivateKey(d)
Point {
  px: {
    c0: 2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198n,
    c1: 3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385n
  },
  py: {
    c0: 1091225907871541971099156319884496768994573942435479350626538810779395867698340173817993335860569537230377198516854n,
    c1: 2595333569697813473526172873449378383222805423158275723970303644390236295775373772592557270513291928304374795997738n
  },
  pz: { c0: 1n, c1: 0n }
}
```

Then using Python to confirm the endianness:

```
$ python
>>> import base64
>>> int.from_bytes(base64.urlsafe_b64decode("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4="), 'big')
15386317252953966243068698475713215110548824422523943950150305918933148006670
>>> int.from_bytes(base64.urlsafe_b64decode("IgRY0ektXHs-heJjAsVTRaMI6OL3W_kYuzGaid2yeQ4="), 'little')
6547402858493071900195431625252553980600119178862449015842830395755105616930
```

So we see that at least the [@noble/curves](https://github.com/paulmillr/noble-curves) implementation uses big-endian encoding of `d`. I haven't compared with other implementations yet, but big-endian is more consistent with the point encoding, since the coordinates (`x.c1` and `x.c0`) are also encoded in big-endian:

```
$ python
>>> x = base64.urlsafe_b64decode("t2IMOefTklK0-lLsfbbnHeik-ax-x-aZVTcXe9heYDme1PhsyDtUzivz-oTg-ZDJDnmbAW1GJwN65w3I0U7q-PgauHW8RoGzc_l7Ac2QBDyKhbfaQ_X0DhqtkAwxXiC2")
>>> int.from_bytes(bytes([x[0] & 0x1f, *x[1:48]]), 'big')
3598972944260337363673575885899865514460004818886439579786207282660602715347404503089745570292800239096749014749385
>>> int.from_bytes(x[48:], 'big')
2227909706099584794832331612878483034628260892607218182073902863691499275449273912364322088077970875819568341590198
```

So I think the definition of `"d"` should change to use big-endian encoding there too.

Author: Changed the description of "d" to big-endian.

}
```

Another example of a different JWK for the BLS12381 curve where the public key is in the G2 subgroup.

```
{
  "kty": "OKP",
  "crv": "BLS12381G2",
  "x": "TBD THIS IS NOW WRONG. REPLACE IT. XLd7zzMpFLeLdr-QQMJ2NPebFkw-h
CyWIj8dwP9uvuu9c9WZrVWvQnY2n7xQHF-BM-Mx6747t_ZJCu809fskA3ca3TVr7",

>   "d": "TBD THIS IS NOW WRONG. REPLACE IT. VaI0tNhs"

Reviewer 2: ```suggestion
  "x": "rgPyS3U6AR_cOuRC1LSxNmw5HfLlAurrRdFGjwyrZihZyeIoqMXQhr7aGM0eo1iS
E2WGstFUjZShhfoF2mK1D61CXFyOAoM-Qt6e7NZjsN9LKZNpGNHnFttuMrlBvk6v",
  "d": "O4x7NVtBvaLb-SZxGe3CKzl7obmSD47aLmReKdKQ_v8"
```

Reviewer 3: `x` and `d` are consistent in my testing.

{
  1: 1,
  -1: 14,
  -2: h'TBD THIS IS NOW WRONG. REPLACE IT. CDAF47031C5D007A32AF743779A14
C4664C602825810FDC71D3E03510A423015A595EA2A0A900D5EDE9FEF8B19A0E35D0E298
43A0622A7CB13EDC3A001BDB12E45B16B60274D28DFBA48253B34D21',

>   -4: h'TBD THIS IS NOW WRONG. REPLACE IT. 50F5858AAFD56EC038739E96F834F'

Reviewer 2: ```suggestion
    -2: h'89bf2eb011792886dc332582aaa7b2f2cd0e0428b3ed9b0f7fdfb3db1cb6c785
ccc3a2db601a3b9fc3b27471f3705cba15be70d37d81bd7e244af2b177c26892fb5770
def9c321d7055b2635cf74b9ce8e300c3d052304a03bf42002fd395b22',
  -4: h'60a16f19075486bf6fda27796b91c9149992d0ffa623b0be8c5e1b413e76d481'
```

Reviewer 3: `x` and `d` are consistent in my testing. Would it make sense for this to be the same key as the corresponding JWK example, though?

{
  1: 1,
  -1: 14,
  -2: h'TBD THIS IS NOW WRONG. REPLACE IT. 49E2228CAA0775A052AE6F1AE5737
DCC66A8C0F0B65FAC070D85B0217CC6C0CE5DB1C27767EF73F2E8355B5F02DABA2B81054
8FC8F0BE26C329375B95B6B6D844CE981CED5934709E57ACA986F927',

>   -4: h'TBD THIS IS NOW WRONG. REPLACE IT. 7305DEEA7631E23D1A27C46BD9F62'

Reviewer 2: ```suggestion
  -2: h'b03ea8b181c4a4d3c5927a9ae3ad98587566b0dba9008240f5d54db7b19ef15c09
2e8baf88394a3fb52f063a6f56488a18e58f81d5ee1bf6c5146adcb1f43d746791322ee275
b022de06ed8b3f8ab56e53a1f6943662b3df528a5640a03eec6a',
  -4: h'3fe7a731678b27201792c20fb11fd91130edc405b8647cfd1ce4598e7186680d'
```

Reviewer 3: `x` and `d` are consistent in my testing. Would it make sense for this to be the same key as the corresponding JWK example, though?
